File: Products/ZenUI3/browser/resources/js/zenoss/service.js
/*
###########################################################################
#
# This program is part of Zenoss Core, an open source monitoring platform.
# Copyright (C) 2009, Zenoss Inc.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 as published by
# the Free Software Foundation.
#
# For complete information please visit: http://www.zenoss.com/oss/
#
###########################################################################
*/
Ext.ns('Zenoss.ui.Service');

Ext.onReady( function() {

    /**********************************************************************
     *
     * Tree Functionality
     *
     */

    /**
     * Callback handler for click events coming from the tree
     * @param {Ext.tree.TreeNode} node Node clicked upon
     */
    function treeClickHandler(node) {
        // Change context
        setPanelsDisabled(false);
    }

    /**
     * Initialize the tree.
     */
    function treeInitialize() {
        Ext.getCmp('master_panel').add({
            xtype: 'HierarchyTreePanel',
            id: 'serviceTree',
            searchField: true,
            autoScroll: true,
            directFn: Zenoss.remote.ServiceRouter.getTree,
            root: 'Services',
            rootuid: '/zport/dmd/Services',
            listeners: {click: treeClickHandler}
        });
    }

    /**********************************************************************
     *
     * CardPanel Functionality
     *
     */

    /**
     * Initialize the Card Panel
     */
    function cardPanelInitialize() {
        Ext.getCmp('bottom_detail_panel').add({
            xtype: 'ContextCardButtonPanel',
            id: 'itCardButtonPanel',
            disabled: true,
            items: [ { xtype: 'SimpleDeviceGridPanel',
                       buttonTitle: _t('Devices'),
                       iconCls: 'devprobs'
                     },
                     { xtype: 'SimpleEventGridPanel',
                       buttonTitle: _t('Events'),
                       iconCls: 'events'
                     }
            ]
        });
    }

    /**********************************************************************
     *
     * Service Panel Functionality
     *
     */

    /**
     * Enables or disables all data entry panels on the screen.
     * @param {boolean} disabled Whether to disable or not
     */
    function setPanelsDisabled(disabled) {
        Ext.getCmp('serviceForm').setDisabled(disabled);
        Ext.getCmp('itCardButtonPanel').setDisabled(disabled);
    }

    /**
     * Enables or disables monitoring options based on inheritance
     * @param {boolean} disabled Whether to disable or not
     */
    function setMonitoringDisabled(disabled) {
        Ext.getCmp('monitorCheckbox').setDisabled(disabled);
        Ext.getCmp('eventSeverityCombo').setDisabled(disabled);
    }

    /**
     * Handles the acquiredCheckbox check events.  If unchecked, it will
     * load its own values, but if checked, it will load its parent's
     * values instead.
     * @param {Ext.form.Checkbox} checkbox The checkbox itself
     * @param {boolean} checked The value of the checkbox as a boolean
     */
    function acquiredCheckboxHandler(checkbox, checked) {
        setMonitoringDisabled(checked);
        var router = Zenoss.remote.ServiceRouter;
        var serviceTree = Ext.getCmp('serviceTree');
        var selectionModel = serviceTree.getSelectionModel();
        var selectedNode = selectionModel.getSelectedNode();

        var uid;
        if (checked && selectedNode.parentNode !== null) {
            uid = selectedNode.parentNode.attributes.uid;
        } else {
            uid = selectedNode.attributes.uid;
        }

        var callback = function(provider, response) {
            var info = response.result.data;
            Ext.getCmp('monitorCheckbox').setValue(info.monitor);
            Ext.getCmp('eventSeverityCombo').setValue(info.eventSeverity);
        };

        router.getInfo({uid: uid, keys: ['monitor', 'eventSeverity']}, callback);
    }

    /**
     * Form definition variables
     */
    var nameTextField = {
        xtype: 'textfield',
        fieldLabel: _t('Name'),
        name: 'name',
        allowBlank: false,
        width: "100%"
    }

    var descriptionTextField = {
        xtype: 'textfield',
        fieldLabel: _t('Description'),
        name: 'description',
        width: "100%"
    }

    var serviceKeysTextField = {
        xtype: 'textfield',
        fieldLabel: _t('Service Keys'),
        name: 'serviceKeys',
        width: "100%"
    }

    var portTextField = {
        xtype: 'textfield',
        fieldLabel: _t('Port'),
        name: 'port',
        width: "100%"
    }

    var acquiredCheckbox = {
        xtype: 'checkbox',
        id: 'acquiredCheckbox',
        fieldLabel: _t('Inherited'),
        name: 'isMonitoringAcquired',
        handler: acquiredCheckboxHandler
    }

    var monitorCheckbox = {
        xtype: 'checkbox',
        id: 'monitorCheckbox',
        fieldLabel: _t('Enabled'),
        name: 'monitor'
    }

    var eventSeverityCombo = {
        xtype: 'combo',
        id: 'eventSeverityCombo',
        fieldLabel: _t('Event Severity'),
        name: 'eventSeverity',
        triggerAction: 'all',
        mode: 'local',
        valueField: 'severityId',
        displayField: 'severityText',
        store: new Ext.data.ArrayStore({
            fields: ['severityId', 'severityText'],
            data: Zenoss.env.SEVERITIES.slice(0, 5)
        })
    }

    var monitoringFieldSet = {
        xtype: 'ColumnFieldSet',
        title: _t('Monitoring'),
        __inner_items__: [
            {
                items: acquiredCheckbox
            }, {
                items: monitorCheckbox,
                bodyStyle: 'padding-left: 15px'
            }, {
                items: eventSeverityCombo,
                bodyStyle: 'padding-left: 15px'
            }
        ]
    }

    var formItems = {
        layout: 'column',
        border: false,
        defaults: {
            layout: 'form',
            border: false,
            bodyStyle: 'padding: 15px',
            labelSeparator: ' ',
            columnWidth: 0.5
        },
        items: [
            {items: [nameTextField,
                     descriptionTextField ]},
            {items: [serviceKeysTextField,
                     portTextField ]},
            {items: [monitoringFieldSet]}
        ]
    }

    var formConfig = {
        xtype: 'form',
        id: 'serviceForm',
        paramsAsHash: true,
        items: formItems,
        border: false,
        labelAlign: 'top',
        autoScroll: true,
        disabled: true,
        api: {
            load: Zenoss.remote.ServiceRouter.getInfo,
            submit: Zenoss.remote.ServiceRouter.submitInfo
        },
        tbar: [
               {
                   xtype: 'button',
                   id: 'Save-button',
                   text: _t('Save'),
                   iconCls: 'save',
                   handler: function(button, event) {
                           var serviceTree = Ext.getCmp('serviceTree');
                           var selectionModel = serviceTree.getSelectionModel();
                           var selectedNode = selectionModel.getSelectedNode();
                           var nameTextField = Ext.getCmp('nameTextField');
                           selectedNode.attributes.text.text = nameTextField.getValue();
                           selectedNode.setText(selectedNode.attributes.text);
                           var form = Ext.getCmp('serviceForm').getForm();
                           var params = Ext.apply({id: selectedNode.id}, form.getValues());
                           form.api.submit(params);
                       }
               }
        ]
    }

    /**
     * @class Zenoss.ui.Service.ServiceFormPanel
     * @extends Ext.form.FormPanel
     * The form panel that displays information about a service organizer or class
     * @constructor
     */
    Zenoss.ui.Service.ServiceFormPanel = Ext.extend(Ext.form.FormPanel, {

        constructor: function(config) {
            Ext.apply(config, formConfig);
            Zenoss.ui.Service.ServiceFormPanel.superclass.constructor.call(this, config);
            //this.on('actioncomplete', actioncompleteHandler);
        }

    });

    Ext.reg('ServiceFormPanel', Zenoss.ui.Service.ServiceFormPanel);

    cardPanelInitialize();
    treeInitialize();

    var serviceForm = new Zenoss.ui.Service.ServiceFormPanel({});


    // place the form in the top right
    Ext.getCmp('top_detail_panel').add(serviceForm);

    serviceForm.getForm().load({params:{uid: 'Services'}});



}); // Ext.onReady

File: Products/Zuul/routers/service.py
###########################################################################
#
# This program is part of Zenoss Core, an open source monitoring platform.
# Copyright (C) 2009, Zenoss Inc.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License version 2 as published by
# the Free Software Foundation.
#
# For complete information please visit: http://www.zenoss.com/oss/
#
###########################################################################

from Products import Zuul
from Products.Zuul.routers import TreeRouter
from Products.Zuul.decorators import require
from Products.ZenUtils.Ext import DirectResponse
from Products.ZenUtils.jsonutils import unjson

class ServiceRouter(TreeRouter):

    def __init__(self, context, request):
        self.api = Zuul.getFacade('service')
        self.context = context
        self.request = request
        super(ServiceRouter, self).__init__(context, request)

    def _getFacade(self):
        return self.api

    def _canDeleteUid(self,uid):
        # check the number of levels deep it is
        levels = len(uid.split('/'))
        return levels > 5

    @require('Manage DMD')
    def addClass(self, contextUid, id, posQuery=None):
        newUid = self.api.addClass(contextUid, id)
        if isinstance(posQuery.get('params'), basestring):
            posQuery['params'] = unjson(posQuery['params'])
        result = self.api.getList(**posQuery)
        for count, serviceInfo in enumerate(result['serviceInfos']):
            if serviceInfo.uid == newUid:
                newIndex = count
                break
        else:
            raise Exception('The new service was added, but the system was '
                            'unable to add it to the list.')
        return DirectResponse(newIndex=newIndex)

    def query(self, limit=None, start=None, sort=None, dir=None, params=None,
              history=False, uid=None, criteria=()):
        if uid is None:
            uid = self.context

        if isinstance(params, basestring):
            params = unjson(params)

        services = self.api.getList(limit, start, sort, dir, params, uid,
                                  criteria)

        disabled = not Zuul.checkPermission('Manage DMD')

        data = Zuul.marshal(services['serviceInfos'])
        return DirectResponse(services=data, totalCount=services['total'],
                              hash=services['hash'], disabled=disabled)

    def getTree(self, id):
        tree = self.api.getTree(id)
        data = Zuul.marshal(tree)
        return [data]

    def getOrganizerTree(self, id):
        tree = self.api.getOrganizerTree(id)
        data = Zuul.marshal(tree)
        return [data]

    def getInfo(self, uid, keys=None):
        service = self.api.getInfo(uid)
        data = Zuul.marshal(service, keys)
        disabled = not Zuul.checkPermission('Manage DMD')
        return DirectResponse.succeed(data=data, disabled=disabled)

    def getParentInfo(self, uid, keys=None):
        service = self.api.getParentInfo(uid)
        data = Zuul.marshal(service, keys)
        disabled = not Zuul.checkPermission('Manage DMD')
        return DirectResponse.succeed(data=data, disabled=disabled)

    def getClassNames(self, uid=None, query=None):
        data = self.api.getClassNames(uid, query)
        return DirectResponse.succeed(data=data)

    @require('Manage DMD')
    def setInfo(self, **data):
        service = self.api.getInfo(data['uid'])
        if data.has_key('serviceKeys') and isinstance(data['serviceKeys'], str):
            data['serviceKeys'] = \
                    tuple([l.strip() for l in data['serviceKeys'].split(',')])
        Zuul.unmarshal(data, service)
        return DirectResponse.succeed()

    def getDevices(self, uid, start=0, params=None, limit=50, sort='device',
                   dir='ASC'):
        if isinstance(params, basestring):
            params = unjson(params)
        devices = self.api.getDevices(uid, start=start, params=params,
                                      limit=limit, sort=sort, dir=dir)
        data = Zuul.marshal(devices)
        return DirectResponse.succeed(data=data)

    def getEvents(self, uid, start=0, params=None, limit=50, sort='device',
                   dir='ASC'):
        if isinstance(params, basestring):
            params = unjson(params)
        events = self.api.getEvents(uid, start=start, params=params,
                                    limit=limit, sort=sort, dir=dir)
        data = Zuul.marshal(events)
        return DirectResponse.succeed(data=data)

    def getInstances(self, uid, start=0, params=None, limit=50, sort='name',
                   dir='ASC'):
        if isinstance(params, basestring):
            params = unjson(params)
        instances = self.api.getInstances(uid, start=start, params=params,
                                          limit=limit, sort=sort, dir=dir)

        keys = ['description', 'device', 'locking', 'monitored', 'name',
                'status', 'uid']
        data = Zuul.marshal(instances, keys)
        return DirectResponse.succeed(data=data, totalCount=instances.total)

    @require('Manage DMD')
    def moveServices(self, sourceUids, targetUid):
        self.api.moveServices(sourceUids, targetUid)
        return DirectResponse.succeed()

    def getUnmonitoredStartModes(self, uid):
        data = self.api.getUnmonitoredStartModes(uid)
        return DirectResponse.succeed(data=Zuul.marshal(data))

    def getMonitoredStartModes(self, uid):
        data = self.api.getMonitoredStartModes(uid)
        return DirectResponse.succeed(data=Zuul.marshal(data))
